### ISSUE TITLE
Spike: how does Python typically handle call outs to other languages?

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#9015]
From the pyChapel implementation, I know that one strategy involves a decorator on the function which uses the other programming language.  The code being imported could then be either written inline using a documentation comment or in a separate file that was specified by the decorator.

Is this the typical strategy?  Or is there a better one?  This spike will involve investigating known Python interoperation (C and numpy, for example), as well as seeking out other languages that Python interoperates with.  It will also involve looking at the implementation of pyChapel, to see how the strategy employed there lines up with these examples (and if there is justification for reusing or revamping that implementation)

**Comment by buddha314** [comment]
I would like to suggest you research how Python can take a variety of complex types in and out of other languages.  One limitation with pyChapel now (or last I looked) was the restriction to primitives.

For instance, we are building complex extensions of the matrix class and I'd love to send them into Python.  If you come across any tips, please share.

**Comment by lydia-duncan** [comment]
There is an extensive list of Python interoperability [here](https://wiki.python.org/moin/IntegratingPythonWithOtherLanguages), both calling from Python and calling to Python.  For now, I'm focusing on calling from Python - several strategies do support both and those might be interesting to consider (possibly for non-performance critical libraries that we don't yet supply), but I think the priority should remain on Python-to-Chapel.

**Comment by lydia-duncan** [comment]
It would also be good to investigate how to get on this list once our support is production ready

**Comment by bradcray** [comment]
> several strategies do support both and those might be interesting to consider

I agree with your statement on which direction should be the priority, but I think we'd ultimately like to support both directions if we could, so shouldn't do anything to preclude that possibility.

In Monday's meeting, it sounded like @mppf had some experience and thoughts here -- hope he's had the chance to share them.

**Comment by lydia-duncan** [comment]
Yup!  We were able to talk yesterday
